scriptworker/cot/verify.py

```diff
@@ -0,0 +1,1141 @@
+#!/usr/bin/env python
+"""Chain of Trust artifact verification.
+
+Attributes:
+    log (logging.Logger): the log object for this module.
+"""
+import asyncio
+from contextlib import contextmanager
+from copy import deepcopy
+from frozendict import frozendict
+import logging
+import os
+import pprint
+import re
+import shlex
+from urllib.parse import unquote, urljoin, urlparse
+from scriptworker.constants import DEFAULT_CONFIG
+from scriptworker.exceptions import CoTError, ScriptWorkerGPGException
+from scriptworker.gpg import get_body, GPG
+from scriptworker.task import download_artifacts, get_decision_task_id, get_worker_type, get_task_id
+from scriptworker.utils import format_json, get_hash, load_json, makedirs, match_url_regex, raise_future_exceptions
+from taskcluster.exceptions import TaskclusterFailure
+
+log = logging.getLogger(__name__)
+
+
+# ChainOfTrust {{{1
+class ChainOfTrust(object):
+    """The master Chain of Trust, tracking all the various `LinkOfTrust`s.
+
+    Attributes:
+        context (scriptworker.context.Context): the scriptworker context
+        cot_dir (str): the local path containing this link's artifacts
+        decision_task_id (str): the task_id of self.task's decision task
+        links (list): the list of `LinkOfTrust`s
+        name (str): the name of the task (e.g., signing.decision)
+        task_id (str): the taskId of the task
+        task_type (str): the task type of the task (e.g., decision, build)
+        worker_impl (str): the taskcluster worker class (e.g., docker-worker) of the task
+    """
+    def __init__(self, context, name, task_id=None):
+        self.name = name
+        self.task_type = guess_task_type(name)
+        self.context = context
+        self.task_id = task_id or get_task_id(context.claim_task)
+        self.task = context.task
+        self.worker_impl = guess_worker_impl(self)  # this should be scriptworker
+        self.decision_task_id = get_decision_task_id(self.task)
+        self.links = []
+
+    def dependent_task_ids(self):
```

Comment:
It doesn't ever shrink, but it's only the links that allow us to trace our task back to the tree.  So signing will have a much shorter chain than, say, balrog.

If we want to test the entire chain of trust, we can do so after the graph finishes with an external audit task.  We can't test the chain of trust artifact from a link further down the chain while the graph is still running, since those tasks haven't run yet.